src/wasm/literal.cpp

```diff
@@ -241,23 +226,47 @@ bool Literal::operator==(const Literal& other) const {
   if (type != other.type) {
     return false;
   }
-  if (type == Type::none) {
-    return true;
-  }
-  if (isNull() || other.isNull()) {
-    return isNull() == other.isNull();
-  }
-  if (type.isFunction()) {
-    return func == other.func;
-  }
-  if (type.isException()) {
-    assert(exn != nullptr && other.exn != nullptr);
-    return *exn == *other.exn;
+  if (type.isBasic()) {
+    switch (type.getBasic()) {
+      case Type::none:
+        return true; // special voided literal
+      case Type::i32:
+      case Type::f32:
+        return i32 == other.i32;
+      case Type::i64:
+      case Type::f64:
+        return i64 == other.i64;
+      case Type::v128:
+        return memcmp(v128, other.v128, 16) == 0;
+      case Type::funcref:
+      case Type::externref:
+      case Type::exnref:
+      case Type::anyref:
+      case Type::eqref:
+      case Type::i31ref:
+        goto is_ref;
```

Comment:
You could avoid this `goto` by defining a `compareRef` lambda up above and calling it here and below. Not that that solution would necessarily be any simpler than the goto, but perhaps it would be less surprising to the goto-wary reader. Same with the one up in `hash`.

Reply:
Seemed like a fine use case for `goto` to keep the code concise, but if that's not deemed good practice a lambda will probably be best, yeah.